### ISSUE TITLE
C/C++ API clarification on accessing arbitrary nodes

### ISSUE THREAD

**Issue by JanSKowalski** [microsoft/onnxruntime#7350]
**What is the recommended approach to accessing the values of an arbitrary node (through C/C++)?**

- Does onnxruntime_cxx_api.h define access for more than the input and output nodes? 

- Could an example function call be provided if node weight access is supported?

I'm not sure how/if the functions GetOverridableInitializer*() are involved. My model seems to have 0 nodes accessible through GetOverridableInitializerCount(), whereas GetInput*() is able to recognize the correct input node.

If it makes a difference, session.Run() on CPU will not be used. The weights of the ONNX model get passed to an FPGA for inference.

Thank you for your time.

**System Information**
I am using this version of the library on ubuntu 20.04:
https://github.com/microsoft/onnxruntime/releases/download/v1.4.0/onnxruntime-linux-x64-1.4.0.tgz

**Additional context - The following resources and issues are relevant:**
https://github.com/onnx/onnx/issues/1425
https://github.com/microsoft/onnxruntime/issues/5560
[CXX_Api_Sample.cpp](https://github.com/microsoft/onnxruntime/blob/master/csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/CXX_Api_Sample.cpp)

[onnxruntime_c_api.h](https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/session/onnxruntime_c_api.h)
[onnxruntime_cxx_api.h](https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/session/onnxruntime_cxx_api.h)
[onnxruntime_cxx_inline.h](https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/session/onnxruntime_cxx_inline.h)

**Comment by hariharans29** [comment]
Not sure if I understood your question completely but I ll give it a shot - 

1) Does onnxruntime_cxx_api.h define access for more than the input and output nodes?
The API doesn't offer access to nodes (not even the "input" and "output" nodes)- it allows you to pass in inputs that may be consumed by these "input" nodes and allows you to get access to the outputs of these "output" nodes you are referring to.

2) Could an example function call be provided if node weight access is supported?
Again, we don't offer access to individual nodes nor their weigths. You can explore the ONNX repo and it offers APIs to construct a model, perform surgery on weights of nodes, append nodes to a model, etc. As a rule of thumb, the ONNX repo provides tools to play around with an ONNX model itself whereas ORT provides a hardware accelerated library for you to run the afore-mentioned model on an environment of your choice.

3) My model seems to have 0 nodes accessible through GetOverridableInitializerCount()
GetOverridableInitializerCount gets counts of initializers (weights) that are also part of the input list. If your initializer is also part of the input list - it is overridable in the sense that you may feed it "weights" of your choice circumventing any values currently held by the initializer. But there are perf risks involved when initializers are part of the input lists (ORT cannot treat them as "constants" and may be unable to do certain optimizations). For more recent models, if you find an initializer in the input list, it is most probably a bug - unless the person who added it really knew what they were doing and understood the implications.
TL;DR - Your model having 0 overridable initializers is nothing to worry about.

**Comment by JanSKowalski** [comment]
I see. Thank you kindly.
